### ISSUE TITLE
[Bug]: tools/check_typehints.py reports errors in type stubs

### ISSUE THREAD

**Issue by dstansby** [matplotlib/matplotlib#27593]
### Bug summary

Running `python tools/check_typehints.py` reports errors. Looking at the source code, these look like real errors. Shouldn't these be caught by our CI?

### Code for reproduction

```python
n/a
```

### Actual outcome

```
 python tools/check_typehints.py 
lib/matplotlib/figure.py::FigureBase::get_tightbbox args differ for get_tightbbox: ['self', 'renderer', 'bbox_extra_artists'] vs ['self', 'renderer']
lib/matplotlib/figure.py::FigureBase::get_tightbbox get_tightbbox kwonlyargs differ: [] vs ['bbox_extra_artists']
lib/matplotlib/rcsetup.py: {'defaultParams'} in stubs and not source
lib/matplotlib/image.py: {'HAMMING', 'LANCZOS', 'HERMITE', 'SPLINE16', 'BESSEL', 'QUADRIC', 'BLACKMAN', 'SINC', 'resample', 'GAUSSIAN', 'KAISER', 'BICUBIC', 'HANNING', 'MITCHELL', 'NEAREST', 'BILINEAR', 'SPLINE36', 'CATROM'} in stubs and not source
lib/matplotlib/colors.py: {'CSS4_COLORS', 'BASE_COLORS', 'XKCD_COLORS', 'TABLEAU_COLORS'} in stubs and not source
lib/matplotlib/axes/_axes.py::Axes::pcolorfast pcolorfast kwonlyargs differ: ['alpha', 'norm', 'cmap', 'vmin', 'vmax'] vs ['alpha', 'norm', 'cmap', 'vmin', 'vmax', 'data']
lib/matplotlib/axes/_axes.py::Axes::plot_date plot_date kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::quiver quiver kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::violinplot violinplot kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::ecdf ecdf kwonlyargs differ: ['complementary', 'orientation', 'compress'] vs ['complementary', 'orientation', 'compress', 'data']
lib/matplotlib/axes/_axes.py::Axes::hist hist kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::stem stem kwonlyargs differ: ['linefmt', 'markerfmt', 'basefmt', 'bottom', 'label', 'orientation'] vs ['linefmt', 'markerfmt', 'basefmt', 'bottom', 'label', 'orientation', 'data']
lib/matplotlib/axes/_axes.py::Axes::scatter scatter kwonlyargs differ: ['edgecolors', 'plotnonfinite'] vs ['edgecolors', 'plotnonfinite', 'data']
lib/matplotlib/axes/_axes.py::Axes::specgram specgram kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::pcolormesh pcolormesh kwonlyargs differ: ['alpha', 'norm', 'cmap', 'vmin', 'vmax', 'shading', 'antialiased'] vs ['alpha', 'norm', 'cmap', 'vmin', 'vmax', 'shading', 'antialiased', 'data']
lib/matplotlib/axes/_axes.py::Axes::cohere cohere kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::phase_spectrum phase_spectrum kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::magnitude_spectrum magnitude_spectrum kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::broken_barh broken_barh kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::acorr acorr kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::eventplot eventplot kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::fill_between fill_between kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::psd psd kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::stairs stairs kwonlyargs differ: ['orientation', 'baseline', 'fill'] vs ['orientation', 'baseline', 'fill', 'data']
lib/matplotlib/axes/_axes.py::Axes::pie pie kwonlyargs differ: ['normalize', 'hatch'] vs ['normalize', 'hatch', 'data']
lib/matplotlib/axes/_axes.py::Axes::errorbar errorbar kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::hlines hlines kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::pcolor pcolor kwonlyargs differ: ['shading', 'alpha', 'norm', 'cmap', 'vmin', 'vmax'] vs ['shading', 'alpha', 'norm', 'cmap', 'vmin', 'vmax', 'data']
lib/matplotlib/axes/_axes.py::Axes::contour contour kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::boxplot boxplot kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::barbs barbs kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::imshow imshow kwonlyargs differ: ['aspect', 'interpolation', 'alpha', 'vmin', 'vmax', 'origin', 'extent', 'interpolation_stage', 'filternorm', 'filterrad', 'resample', 'url'] vs ['aspect', 'interpolation', 'alpha', 'vmin', 'vmax', 'origin', 'extent', 'interpolation_stage', 'filternorm', 'filterrad', 'resample', 'url', 'data']
lib/matplotlib/axes/_axes.py::Axes::bar bar kwonlyargs differ: ['align'] vs ['align', 'data']
lib/matplotlib/axes/_axes.py::Axes::xcorr xcorr kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::contourf contourf kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::vlines vlines kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::angle_spectrum angle_spectrum kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::fill_betweenx fill_betweenx kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::hexbin hexbin kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::hist2d hist2d kwonlyargs differ: [] vs ['data']
lib/matplotlib/axes/_axes.py::Axes::csd csd kwonlyargs differ: [] vs ['data']
lib/matplotlib/_api/deprecation.py: {'DeprecationKwargs', 'NamedDeprecationKwargs'} in stubs and not source
lib/matplotlib/_api/deprecation.py::deprecated deprecated kwonlyargs differ: ['message', 'name', 'alternative', 'pending', 'obj_type', 'addendum', 'removal'] vs []
lib/matplotlib/_api/deprecation.py::deprecated deprecated varkwarg differ: None vs kwargs
lib/matplotlib/_api/deprecation.py::deprecate_privatize_attribute::__init__ args differ for __init__: ['self'] vs ['self', 'since']
lib/matplotlib/_api/deprecation.py::deprecate_privatize_attribute::__init__ __init__ vararg differ: args vs None
lib/matplotlib/_api/deprecation.py::warn_deprecated warn_deprecated kwonlyargs differ: ['message', 'name', 'alternative', 'pending', 'obj_type', 'addendum', 'removal'] vs []
lib/matplotlib/_api/deprecation.py::warn_deprecated warn_deprecated varkwarg differ: None vs kwargs
lib/matplotlib/_api/deprecation.py::deprecate_method_override deprecate_method_override kwonlyargs differ: ['allow_empty'] vs ['allow_empty', 'since']


49 total errors found
```

### Expected outcome

0 errors found

### Matplotlib Version

3.9.0.dev905+gb4fc72bb88.d20240102

### Installation

git checkout

**Comment by tacaswell** [comment]
This is a question for @ksunden .

**Comment by ksunden** [comment]
`check_typehints` is a script I wrote myself prior to gaining a better understanding of `mypy`'s `stubtest`. The latter is what we actually use in CI, but we did not get _rid_ of my check script because it is not quite _the same_.

These were used as a guide, though never made user friendly/able to use e.g. CLI options to narrow checks (though some of that is available by editing the script...)


Of these checks:

All of the ones in `_axes.py` are only that the `data` kwarg is not handled in type hints currently... It introduces type instability that is kind of hard to accurately capture, and I chose not to worry about it in the first pass at writing type hints, and would prefer some way of using the `_preprocess_data` decorator to automatically apply the changes, though I believe the tooling for the kinds of changes introduced is not available at the language level for specifying the effects. Any ideas for handling `data` without being overly burdensome are certainly welcome. I for one am not looking forward to wrapping all of the artist methods with `@overload` manually and adding `| str` to most fields... but that may be necessary.

The flags from `colors.py` and `image.py` is because these are imported from private submodules in the `.py` files, but are available publically from these files. I did not type hint most private behavior initially, so just put those in their public location.

`get_tightbox` I believe is reflecting the API after `make_keyword_only` deprecation period. The docstring for the `check_typehints` script indicates it expects that `make_keyword_only` will be changed on expiry, but either choice is defensible.

The things in `deprecation.py` include some types introduced to represent shared kwargs across several functions, and my script does not handle unpacking those kwargs (but I think stubtest does)

`defaultParams` is an odd one that is placed in that namespace dynamically and is therefore not in the source, but is relatively commonly used.

The `stubtest` configuration references the `ci/mypy-stubtest-allowlist.txt` file, which ignore many flags relating to e.g. private behavior that was never type hinted etc.
I'd be in favor of working to reduce those ignores.

We have an alternate script in `tests/stubtest.py` that _is_ used in CI and does a pass at automatically ignoring our dynamic behavior (e.g. `define_aliases` decorator and `delete_parameter`) and then runs `stubtest` with consistent parameters, specifying our allowlist, etc.
